pootle/core/checks/checker.py

```diff
@@ -0,0 +1,305 @@
+        return units
+
+    def clear_checks(self):
+        QualityCheck.delete_unknown_checks()
+
+    @lru_cache(maxsize=None)
+    def get_checker(self, tp):
+        """Return the site QualityChecker or the QC associated with
+        the a Unit's TP otherwise.
+        """
+        if settings.POOTLE_QUALITY_CHECKER:
+            return import_func(settings.POOTLE_QUALITY_CHECKER)()
+        try:
+            return TranslationProject.objects.get(id=tp).checker
+        except TranslationProject.DoesNotExist:
+            # There seems to be a risk of dangling Stores with no TP
+            return None
+
+    def update(self):
+        """Update/purge all QualityChecks for Units, and expire Store caches.
+        """
+        start = time.time()
+        self.clear_checks()
+        logger.debug(
+            "deleted unknown checks in %s seconds"
+            % (time.time() - start))
+
+        start = time.time()
+        untrans = self.update_untranslated()
+        logger.debug(
+            "deleted %s checks for untranslated units in %s seconds"
+            % (untrans, (time.time() - start)))
+
+        start = time.time()
+        trans = self.update_translated()
+        logger.debug(
+            "updated checks for %s units in %s seconds"
+            % (trans, (time.time() - start)))
+
+        start = time.time()
+        self.update_store_caches()
+        logger.debug(
+            "expired %s Store caches in %s seconds"
+            % (len(self.stores), (time.time() - start)))
+
+    def update_store_caches(self):
+        """After completing QC updates expire caches for affected Stores.
+        """
+        for store in Store.objects.filter(pk__in=self.stores):
+            store.update_dirty_cache()
+
+    def update_translated_unit(self, unit, checker=None):
+        """Update checks for a translated Unit
+        """
+        if checker is None:
+            checker = self.get_checker(unit[tp_key])
+        if checker is None:
+            return False
+        if self.translation_project is not None:
+            # if TP is set then manually add TP.id to the Unit value dict
+            unit[tp_key] = self.translation_project.id
+        unit = CheckableUnit(unit)
+        checker = UnitQualityCheck(
+            checker,
+            unit,
+            self.checks.get(unit.id, {}),
+            self.check_names,
+            self.keep_false_positives)
+        if checker.update():
+            self.stores.add(unit.store)
+            self.units.filter(id=unit.id).update(mtime=timezone.now())
+            return True
+        return False
+
+    def update_translated(self):
+        """Update checks for translated Units
+        """
+        unit_fields = [
+            "id", "source_f", "target_f", "locations", "store__id"]
+
+        tp_key = "store__translation_project__id"
+        if self.translation_project is None:
+            unit_fields.append(tp_key)
```

Comment:
it does get reused but there is an error in the code - which i have fixed - but waiting to push